### PR TITLE
Cleanup csharp API SessionOptions and RunOptions to be consistent with other APIs

### PR DESCRIPTION
**Pull request by shahasad**
**Description**: 
- Made the SessionOptions and RunOptions containers of read/write properties, similar to Python API.
- Added missing options. 
- Removed the Default singleton for SessionOptions and RunOptions due to unclear semantics (user might change the default, what happens if use disposes the default, etc.)
- exposed the InferenceSession.Run() API with RunOptions
- exposed the LogLevel enum

**Motivation and Context**
- Why is this change required? What problem does it solve?
To be consistent across different API bindings, before we ship 1.0 release

- If it fixes an open issue, please link to the issue here.

** Note: There are breaking changes from prior API **

### CODE REVIEW

**Comment by pranavsharma** [comment]
Looks mostly ok. I'm not an expert in C#. May be @jignparm can help review as well?

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 1]
```diff
@@ -82,32 +160,51 @@ private void CanRunInferenceOnAModel(uint graphOptimizationLevel, bool disableSe
...
+                    IReadOnlyCollection<string> outputNames = session.OutputMetadata.Keys.ToList();
 
-                    float[] expectedOutput = LoadTensorFromFile(@"bench.expected_out");
-                    // validate the results
-                    foreach (var r in results)
+                    using (var results = session.Run(container, outputNames, runOptions))  // results is an IReadOnlyList<NamedOnnxValue> container
                     {
-                        Assert.Equal("softmaxout_1", r.Name);
+                        validateRunResults(results);
+                    }
```
Comment (jignparm) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs]: Does this run generate a log file that can be tested (i.e. it exists and has non-empty content)? If so, should we test for the log file?

Comment (shahasad) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs]: Currently all logs go to console, and we don't have run/session option to redirect the log output. Only logger override is available at the ENV, which is not exposed in C#. I don't know a good way to look into vstest console output either, and did not want to spend time parsing it.

**Comment by shahasad** [comment]
> Looks mostly ok. I'm not an expert in C#. May be @jignparm can help review as well?

Please do check at the API level, the names of the properties/functions are OK.

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs [hunk 2]
```diff
@@ -181,6 +181,37 @@ internal static class NativeMethods
...
+        [DllImport(nativeLib, CharSet = charSet)]
+        public static extern IntPtr /*(OrtStatus*)*/ OrtRunOptionsGetRunTag(IntPtr /* const OrtRunOptions* */options, out IntPtr /* const char** */ runtag);
+
+        // Set a flag so that any running OrtRun* calls that are using this instance of OrtRunOptions
+        // will exit as soon as possible if the flag is true.
+        [DllImport(nativeLib, CharSet = charSet)]
+        public static extern IntPtr /*(OrtStatus*)*/ OrtRunOptionsEnableTerminate(IntPtr /* OrtRunOptions* */ options);
+
+        [DllImport(nativeLib, CharSet = charSet)]
+        public static extern IntPtr /*(OrtStatus*)*/ OrtRunOptionsDisableTerminate(IntPtr /* OrtRunOptions* */ options);
```
Comment (jignparm) [inline csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs]: Any tests that can be put in for OrtRunOptionsEnableTerminate, OrtRunOptionsEnableTerminate, OrtRunOptionsGetRunTag etc?

Comment (jignparm) [reply csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs]: Also,  **VerifyNativeMethodsExist()** needs to be updated for these new API to check against future name changes.

File: csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs [hunk 3]
```diff
@@ -4,117 +4,274 @@
...
         /// Constructs an empty SessionOptions
         /// </summary>
         public SessionOptions()
         {
-            NativeMethods.OrtCreateSessionOptions(out _nativePtr);
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateSessionOptions(out _nativePtr));
+        }
+
+
+        private static SessionOptions MakeSessionOptionWithCpuProvider()
```
Comment (jignparm) [inline csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs]: Seems like **MakeSessionOptionWithCpuProvider()** is not referenced anywhere in the project any more.

For Linux OS, the CheckLibcVersionGreaterThanMinimum() function still needs to be run before an InferenceSession is created, to check if the C runtime libraries on the OS are supported by ORT. Previously this check was run in  **MakeSessionOptionWithCpuProvider()** to guarantee the sequence -- does this check still get run somehow?

**Comment by jignparm** [comment]
LGTM. Putting CheckLibcVersionGreaterThanMinimum() in the constructor will ensure execution.

**Pull request merged**